### ISSUE TITLE
Bitwise assignment doesn't change type

### ISSUE THREAD

**Issue by bugreportuser** [vimeo/psalm#1340]
If `$x` is `false`, using `|= true` should make it `1`.

Example (https://getpsalm.org/r/11805c0b24):
```php
<?php
$x = false;
$x |= rand(0, 1) !== 2;
$x |= true;
$x |= 1;
if ($x) {
    echo $x;
}
```

Example output:
```
Psalm output (using commit a428b34):

ERROR: TypeDoesNotContainType - 6:5 - Found a contradiction when evaluating $x and trying to reconcile type 'false' to !falsy
```

**Comment by muglug** [comment]
I've fixed this for ints - not sure about bools

**Comment by bugreportuser** [comment]
Thanks. Are you willing to add support for bools later or at least a warning?

**Comment by muglug** [comment]
Yeah, going to keep this open because it's a valid use case

**Comment by SignpostMarv** [comment]
Just checking in; is the following example a separate issue to this one? https://psalm.dev/r/222baddf7b

**Comment by bugreportuser** [comment]
@SignpostMarv It's a separate issue but similar to #2139.

**Comment by orklah** [comment]
This has been reworked since and now boolean operations with bitwise operators emits an error. Not sure we can still consider this a valid use case

**Comment by bugreportuser** [comment]
You're right. I've had to avoid this for other tools too, so this issue isn't a priority. It's surprising php hasn't started supporting bitwise operators for booleans since it would be very useful

**Comment by tvdijen** [comment]
This issue still exists, yet the error messages have changed over time;
https://psalm.dev/r/aa2d948c03

**Comment by orklah** [comment]
I'm not sure why solving boolean operations make the strpos operation go without Issue, can you explain that?

**Comment by AndrolGenhald** [comment]
I pushed my attempt [here](https://github.com/vimeo/psalm/compare/master...AndrolGenhald:feature/1340-booleans-in-arithmetic) so you can look at it. The problem is that the `bool` to `int` cast happens before the `PossiblyFalseOperand` is emitted. I hadn't considered to try both allowing the cast and emitting the issue, but I thought the point was more to make Psalm allow this pattern without any issues rather than to still show issues but have the correct types.

Especially after the casting improvements in #7696 I think the best solution for now is just for users to cast when doing this, it will result in correct types and no issues from Psalm.
